docs/proposals/ldap-group-sync.md

```diff
@@ -0,0 +1,189 @@
+}
+
+func (s *LDAPGroupSyncer) SyncAllGroups() (err error) {
+  // use s.cachedMemberLister to determine groups to sync and find members for each
+  // for each group:
+  //    users := s.determineUsers(members)
+  // use s.groupClient to determine which Group to update
+}
+
+func (s *LDAPGroupSyncer) SyncGroups(groupFilters util.StringSet) (err error) {
+  // use s.memberLister to map groupFilter -> members
+  // for each group:
+  //    users := s.determineUsers(members)
+  // use s.groupClient to determine which Group to update
+}
+
+func (s *LDAPGroupSyncer) determineUsers(members []*ldap.Entry) (users []*userapi.User, err error) {
+  // use s.identityFactory to map member -> Identity
+  // use s.identityMapper to determine Identity -> user
+}
+
+```
+`LDAPSyncOptions` needs the following schema (TBD) :
+```
+LDAP server information (would this be in the identityprovider?)
+LDAP schema has first class groups?
+common LDAP filter for "all" (groups, users)
+specific filter key (for groups and for users)
+mapping OpenShift Group to specific filter value (either by opaque values or programatically from Group fields)
+mapping LDAP group to specific filter value
+how to map LDAP group attribute specifying a member to specific filter value
+attribute on first-class member that specifies groups or on first-class group that specifies members
+mapping from first-class LDAP group attributes to OpenShift Group fields
+```
+
+##  Determining Which Groups To Sync
+### Determining What LDAP Groups Exist
+If the LDAP schema has first-class groups, then determining which groups exist is simple once the LDAP request for all groups is run. If the LDAP schema encodes groups as attributes for first-class users, then determining the list of group names needs to be done from the result of the LDAP query for all users. This query is cached for the `MemberLister` in this case.
+
+### Forming LDAP Queries for OpenShift `Groups`
+All LDAP requests will contain a conjunction of two filters: a filter that matches all of the LDAP groups or LDAP users that are relevant, and a second filter that, when conjoined with the first matches only the specific LDAP group or set of LDAP members that define a group's membership. An `LDAPFilterGenerator` is used to map from an OpenShift `Group` the filter that will be used to recover it's mirror from the LDAP server.
+
+```go
+// LDAPFilterGenerator generates a filter to match a Group, given a Group
+type LDAPFilterGenerator interface {
+  GenerateFilter(group Group) (string filter, err error)
+}
+```
+
+#### Given a Mapping of OpenShift `Groups` to LDAP Filters
+OpenShift will support users providing opaque LDAP filters along with the other LDAP request fields in order to create an LDAP request. Opaque filters will be indexed using a unique identifier on groups stored as an `Annotation` such as their distinctive name in the LDAP server or value of the memberOf annotation that designates a LDAP user as a member of the group.
```

Comment:
> This is for Schema 2, where, no, there may not be a common way to query.

Imprecise question:  In the schema2 case, do we expect to possible be possible to describe a substitutable query that would accept the name of the group as a parameter?  If so, the query should live in the config and the openshift group only holds the ldap group name.  If not, do we care enough about this case to allow the additional annotation?  It seem unlikely that an admin would get it right.